File: go/indicators/geraldappel/movingaverageconvergencedivergence/movingaverageconvergencedivergence.go
package movingaverageconvergencedivergence

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/core"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// MovingAverageConvergenceDivergence is Gerald Appel's MACD indicator.
//
// MACD is calculated by subtracting the slow moving average from the fast moving average.
// A signal line (moving average of MACD) and histogram (MACD minus signal) are also produced.
//
// The indicator produces three outputs:
//   - MACD: fast MA - slow MA
//   - Signal: MA of the MACD line
//   - Histogram: MACD - Signal
//
// Reference:
//
// Appel, Gerald (2005). Technical Analysis: Power Tools for Active Investors.
type MovingAverageConvergenceDivergence struct {
	mu sync.RWMutex

	fastMA   lineUpdater
	slowMA   lineUpdater
	signalMA lineUpdater

	macdValue      float64
	signalValue    float64
	histogramValue float64
	primed         bool

	// fastDelay is the number of initial samples to skip before feeding the fast MA.
	// This aligns the fast MA's SMA seed window with the slow MA's, matching TaLib's
	// batch algorithm where both MAs start producing output at the same index.
	fastDelay int
	fastCount int

	barFunc   entities.BarFunc
	quoteFunc entities.QuoteFunc
	tradeFunc entities.TradeFunc

	mnemonic string
}

// NewMovingAverageConvergenceDivergence returns an instance of the indicator created using supplied parameters.
//
//nolint:funlen,cyclop
func NewMovingAverageConvergenceDivergence(
	p *MovingAverageConvergenceDivergenceParams,
) (*MovingAverageConvergenceDivergence, error) {
	const (
		invalid           = "invalid moving average convergence divergence parameters"
		fmts              = "%s: %s"
		fmtw              = "%s: %w"
		minLength         = 2
		minSignalLength   = 1
		defaultFastLength = 12
		defaultSlowLength = 26
		defaultSignalLen  = 9
	)

	fastLength := p.FastLength
	if fastLength == 0 {
		fastLength = defaultFastLength
	}

	slowLength := p.SlowLength
	if slowLength == 0 {
		slowLength = defaultSlowLength
	}

	signalLength := p.SignalLength
	if signalLength == 0 {
		signalLength = defaultSignalLen
	}

	if fastLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast length should be greater than 1")
	}

	if slowLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than 1")
	}

	if signalLength < minSignalLength {
		return nil, fmt.Errorf(fmts, invalid, "signal length should be greater than 0")
	}

	// Auto-swap fast/slow if needed (matches TaLib behavior).
	if slowLength < fastLength {
		fastLength, slowLength = slowLength, fastLength
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Default FirstIsAverage to true (TA-Lib compatible: EMA seeded with SMA).
	firstIsAverage := true
	if p.FirstIsAverage != nil {
		firstIsAverage = *p.FirstIsAverage
	}

	var fastMA, slowMA, signalMA lineUpdater

	// Create fast and slow MAs.
	fastMA, err = newMA(p.MovingAverageType, fastLength, firstIsAverage)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	slowMA, err = newMA(p.MovingAverageType, slowLength, firstIsAverage)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Create signal MA.
	signalMA, err = newMA(p.SignalMovingAverageType, signalLength, firstIsAverage)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := buildMnemonic(fastLength, slowLength, signalLength,
		p.MovingAverageType, p.SignalMovingAverageType, bc, qc, tc)

	return &MovingAverageConvergenceDivergence{
		fastMA:         fastMA,
		slowMA:         slowMA,
		signalMA:       signalMA,
		fastDelay:      slowLength - fastLength,
		macdValue:      math.NaN(),
		signalValue:    math.NaN(),
		histogramValue: math.NaN(),
		barFunc:        barFunc,
		quoteFunc:      quoteFunc,
		tradeFunc:      tradeFunc,
		mnemonic:       mnemonic,
	}, nil
}

func newMA(maType MovingAverageType, length int, firstIsAverage bool) (lineUpdater, error) {
	switch maType {
	case SMA:
		return simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: length})
	default: // EMA (zero value)
		return exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         length,
				FirstIsAverage: firstIsAverage,
			})
	}
}

func maLabel(maType MovingAverageType) string {
	if maType == SMA {
		return "SMA"
	}

	return "EMA"
}

func buildMnemonic(
	fastLen, slowLen, signalLen int,
	maType, signalMAType MovingAverageType,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) string {
	// Default: macd(12,26,9) when both MA types are EMA.
	// Non-default: macd(12,26,9,SMA,EMA) showing both types.
	suffix := ""
	if maType != EMA || signalMAType != EMA {
		suffix = fmt.Sprintf(",%s,%s", maLabel(maType), maLabel(signalMAType))
	}

	return fmt.Sprintf("macd(%d,%d,%d%s%s)", fastLen, slowLen, signalLen, suffix,
		core.ComponentTripleMnemonic(bc, qc, tc))
}

// IsPrimed indicates whether the indicator is primed.
func (s *MovingAverageConvergenceDivergence) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *MovingAverageConvergenceDivergence) Metadata() core.Metadata {
	desc := "Moving Average Convergence Divergence " + s.mnemonic

	return core.BuildMetadata(
		core.MovingAverageConvergenceDivergence,
		s.mnemonic,
		desc,
		[]core.OutputText{
			{Mnemonic: s.mnemonic + " macd", Description: desc + " MACD"},
			{Mnemonic: s.mnemonic + " signal", Description: desc + " Signal"},
			{Mnemonic: s.mnemonic + " histogram", Description: desc + " Histogram"},
		},
	)
}

// Update updates the indicator given the next sample value.
// Returns macd, signal, histogram values.
//
//nolint:nonamedreturns
func (s *MovingAverageConvergenceDivergence) Update(sample float64) (macd, signal, histogram float64) {
	nan := math.NaN()

	if math.IsNaN(sample) {
		return nan, nan, nan
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	// Feed the slow MA every sample.
	slow := s.slowMA.Update(sample)

	// Delay the fast MA to align SMA seed windows (matches TaLib batch algorithm).
	var fast float64

	if s.fastCount < s.fastDelay {
		s.fastCount++
		fast = nan
	} else {
		fast = s.fastMA.Update(sample)
	}

	if math.IsNaN(fast) || math.IsNaN(slow) {
		s.macdValue = nan
		s.signalValue = nan
		s.histogramValue = nan

		return nan, nan, nan
	}

	macd = fast - slow
	s.macdValue = macd

	signal = s.signalMA.Update(macd)

	if math.IsNaN(signal) {
		s.signalValue = nan
		s.histogramValue = nan

		return macd, nan, nan
	}

	s.signalValue = signal
	histogram = macd - signal
	s.histogramValue = histogram
	s.primed = s.fastMA.IsPrimed() && s.slowMA.IsPrimed() && s.signalMA.IsPrimed()

	return macd, signal, histogram
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *MovingAverageConvergenceDivergence) UpdateScalar(sample *entities.Scalar) core.Output {
	macd, signal, histogram := s.Update(sample.Value)

	const outputCount = 3

	output := make([]any, outputCount)
	output[0] = entities.Scalar{Time: sample.Time, Value: macd}
	output[1] = entities.Scalar{Time: sample.Time, Value: signal}
	output[2] = entities.Scalar{Time: sample.Time, Value: histogram}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *MovingAverageConvergenceDivergence) UpdateBar(sample *entities.Bar) core.Output {
	v := s.barFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *MovingAverageConvergenceDivergence) UpdateQuote(sample *entities.Quote) core.Output {
	v := s.quoteFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *MovingAverageConvergenceDivergence) UpdateTrade(sample *entities.Trade) core.Output {
	v := s.tradeFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

File: py/indicators/gerald_appel/moving_average_convergence_divergence/moving_average_convergence_divergence.py
"""Moving Average Convergence Divergence (MACD) indicator."""

import math
from typing import List, Any

from ...core.indicator import Indicator
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.identifier import Identifier
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ....entities.bar import Bar
from ....entities.quote import Quote
from ....entities.trade import Trade
from ....entities.scalar import Scalar
from ....entities.bar_component import BarComponent, DEFAULT_BAR_COMPONENT, bar_component_value
from ....entities.quote_component import QuoteComponent, DEFAULT_QUOTE_COMPONENT, quote_component_value
from ....entities.trade_component import TradeComponent, DEFAULT_TRADE_COMPONENT, trade_component_value
from ...common.simple_moving_average.simple_moving_average import SimpleMovingAverage
from ...common.simple_moving_average.params import SimpleMovingAverageParams
from ...common.exponential_moving_average.exponential_moving_average import ExponentialMovingAverage
from ...common.exponential_moving_average.params import ExponentialMovingAverageLengthParams
from .params import MovingAverageConvergenceDivergenceParams, MovingAverageType


def _new_ma(ma_type: MovingAverageType, length: int, first_is_average: bool):
    """Create a moving average updater."""
    if ma_type == MovingAverageType.SMA:
        return SimpleMovingAverage(SimpleMovingAverageParams(length=length))
    else:
        return ExponentialMovingAverage.from_length(
            ExponentialMovingAverageLengthParams(
                length=length, first_is_average=first_is_average))


def _ma_label(ma_type: MovingAverageType) -> str:
    if ma_type == MovingAverageType.SMA:
        return "SMA"
    return "EMA"


class MovingAverageConvergenceDivergence(Indicator):
    """Gerald Appel's MACD indicator.

    MACD is calculated by subtracting the slow moving average from the fast moving average.
    A signal line (moving average of MACD) and histogram (MACD minus signal) are also produced.

    The indicator produces three outputs:
      - MACD: fast MA - slow MA
      - Signal: MA of the MACD line
      - Histogram: MACD - Signal
    """

    def __init__(self, p: MovingAverageConvergenceDivergenceParams) -> None:
        fast_length = p.fast_length
        slow_length = p.slow_length
        signal_length = p.signal_length

        if fast_length < 2:
            raise ValueError(
                "invalid moving average convergence divergence parameters: "
                "fast length should be greater than 1")
        if slow_length < 2:
            raise ValueError(
                "invalid moving average convergence divergence parameters: "
                "slow length should be greater than 1")
        if signal_length < 1:
            raise ValueError(
                "invalid moving average convergence divergence parameters: "
                "signal length should be greater than 0")

        # Auto-swap fast/slow if needed (matches TaLib behavior).
        if slow_length < fast_length:
            fast_length, slow_length = slow_length, fast_length

        bc = p.bar_component if p.bar_component is not None else DEFAULT_BAR_COMPONENT
        qc = p.quote_component if p.quote_component is not None else DEFAULT_QUOTE_COMPONENT
        tc = p.trade_component if p.trade_component is not None else DEFAULT_TRADE_COMPONENT

        self._bar_func = bar_component_value(bc)
        self._quote_func = quote_component_value(qc)
        self._trade_func = trade_component_value(tc)

        # Default FirstIsAverage to true (TA-Lib compatible).
        first_is_average = True if p.first_is_average is None else p.first_is_average

        self._fast_ma = _new_ma(p.moving_average_type, fast_length, first_is_average)
        self._slow_ma = _new_ma(p.moving_average_type, slow_length, first_is_average)
        self._signal_ma = _new_ma(p.signal_moving_average_type, signal_length, first_is_average)

        self._fast_delay = slow_length - fast_length
        self._fast_count = 0

        self._macd_value = math.nan
        self._signal_value = math.nan
        self._histogram_value = math.nan
        self._primed = False

        # Build mnemonic.
        suffix = ""
        if p.moving_average_type != MovingAverageType.EMA or \
                p.signal_moving_average_type != MovingAverageType.EMA:
            suffix = f",{_ma_label(p.moving_average_type)},{_ma_label(p.signal_moving_average_type)}"

        self._mnemonic = f"macd({fast_length},{slow_length},{signal_length}" \
                         f"{suffix}{component_triple_mnemonic(bc, qc, tc)})"

    def is_primed(self) -> bool:
        return self._primed

    def metadata(self) -> Metadata:
        desc = f"Moving Average Convergence Divergence {self._mnemonic}"
        return build_metadata(
            Identifier.MOVING_AVERAGE_CONVERGENCE_DIVERGENCE,
            self._mnemonic,
            desc,
            [
                OutputText(f"{self._mnemonic} macd", f"{desc} MACD"),
                OutputText(f"{self._mnemonic} signal", f"{desc} Signal"),
                OutputText(f"{self._mnemonic} histogram", f"{desc} Histogram"),
            ],
        )

    def update(self, sample: float) -> tuple[float, float, float]:
        """Update with a scalar value. Returns (macd, signal, histogram)."""
        nan = math.nan

        if math.isnan(sample):
            return nan, nan, nan

        # Feed the slow MA every sample.
        slow = self._slow_ma.update(sample)

        # Delay the fast MA to align SMA seed windows.
        if self._fast_count < self._fast_delay:
            self._fast_count += 1
            fast = nan
        else:
            fast = self._fast_ma.update(sample)

        if math.isnan(fast) or math.isnan(slow):
            self._macd_value = nan
            self._signal_value = nan
            self._histogram_value = nan
            return nan, nan, nan

        macd = fast - slow
        self._macd_value = macd

        signal = self._signal_ma.update(macd)

        if math.isnan(signal):
            self._signal_value = nan
            self._histogram_value = nan
            return macd, nan, nan

        self._signal_value = signal
        histogram = macd - signal
        self._histogram_value = histogram
        self._primed = self._fast_ma.is_primed() and self._slow_ma.is_primed() \
            and self._signal_ma.is_primed()

        return macd, signal, histogram

    def update_scalar(self, sample: Scalar) -> List[Any]:
        macd, signal, histogram = self.update(sample.value)
        return [
            Scalar(time=sample.time, value=macd),
            Scalar(time=sample.time, value=signal),
            Scalar(time=sample.time, value=histogram),
        ]

    def update_bar(self, sample: Bar) -> List[Any]:
        v = self._bar_func(sample)
        return self.update_scalar(Scalar(time=sample.time, value=v))

    def update_quote(self, sample: Quote) -> List[Any]:
        v = self._quote_func(sample)
        return self.update_scalar(Scalar(time=sample.time, value=v))

    def update_trade(self, sample: Trade) -> List[Any]:
        v = self._trade_func(sample)
        return self.update_scalar(Scalar(time=sample.time, value=v))
